### DISCUSSION TITLE
How to get proper tree of my collection ?

### DISCUSSION THREAD

**Discussion by GhostvOne** [staudenmeir/laravel-adjacency-list#270] [General]
I have two models and i'm trying to get categories tree attached to the file model

1st model : "File"
2nd model : "Category"

Categories can have parents with "category_parent_id" field.
File can have multiple categories (belongsToMany).

Imagine File contain categories (1,2,3,4 )
Actually, if I take the categories collection from the file model
```php
class File {
 // ...
 public function categories(): BelongsToMany
    {
        return $this->belongsToMany(category::class);
    }
}
```

```php
$file->load('categories');
$file->categories->toTree()
```
It returns me : 
 - Plugins (id : 1)
- - Mods (id : 2)
- - Logs (id : 3)
- - GUI (id : 4)
- Mods (id : 2) (it's a child of plugins, don't need it here)
- Logs (id : 3) (it's a child of plugins, don't need it here)
- GUI (id : 4) (it's a child of plugins, don't need it here)

instead of 
- Plugins (id : 1)
- - Mods (id : 2)
- - Logs (id : 3)
- - GUI (id : 4)

How Can I only have the first occurence, plugins with children ? 

Thanks

**Comment by staudenmeir** [comment]
Hi @GhostvOne,
Your tree contains "duplicates" because `toTree()` can't be applied to all types of `Category` results – only the ones with a `depth` attribute.

We could find a workaround, but its complexity depends on your use case.

> Imagine File contain categories (1,2,3,4 )

Can you describe the data of this relationship in general?
Are files always connected to a category (`1`) and (all of) its descendants (`2`, `3`, `4`)?
Or files can be connected to multiple categories that aren't in the same hierarchy (e.g. to `5`, a sibling of `1`)?

**Reply by GhostvOne** [reply]
hello,
you are right, file can be connected to multiple categories its the own choose of the users

Imagine categories : 

- Plugins (id : 1)
- - Mods (id : 2)
- - Logs (id : 3)
- - GUI (id : 4)
- Cosmetics (id : 5)
- - A (id : 6)
- - B (id : 7) 
- - C (id : 8)
- - - D (id : 9)
- - - E (id : 10)

A file can be attached to : 1 + 2 + 4 and 5 + 6 + 8 + 9
My relations is returning me every ids (from a pivot table)

**Reply by staudenmeir** [reply]
Try this filter:

```php
$tree = $file->categories->toTree();

$ids = $tree->pluck('id');

$tree = $tree->filter(
    fn (Category $category) => !$ids->contains($category->category_parent_id)
)->values();
```
